core/src/main/java/org/apache/iceberg/avro/ValueWriters.java

```diff
@@ -358,8 +380,10 @@ public void write(BigDecimal decimal, Encoder encoder) throws IOException {
     private final int nullIndex;
     private final int valueIndex;
     private final ValueWriter<T> valueWriter;
+    private final Schema.Type type;
+    private long nullValueCount;
 
-    private OptionWriter(int nullIndex, ValueWriter<T> valueWriter) {
+    private OptionWriter(int nullIndex, ValueWriter<T> valueWriter, Schema.Type type) {
```

Comment:
It looks like the only place that `type` is used is in an error message. I'd prefer not to change the signature here just to print a type.

Reply:
Oh I think this `type` is also used for checking `isMetricSupportedType/supportsMetrics` so we will still need it.

Reply:
I think that we should not pass the type here. Instead, let's have a `MetricsOptionWriter` and a non-metrics version and choose which one to use ahead of time. That way, we don't keep a count that won't be used. There's no need to check whether the null count should be used when `metrics` is called. That can be determined ahead of time.

If we do that, then there is no need to pass the type here or into the `option` method. We can either add a boolean (`collectMetrics`) or have a separate factory method. I think that would be a bit cleaner.